python/ray/util/sgd/v2/migration/sgd-migration-guide.md

```diff
@@ -0,0 +1,224 @@
+
+from ray.sgd import Trainer
+
+trainer = Trainer(backend="torch", num_workers=4)
+trainer.start()
+results = trainer.run(train_func_distributed)
+trainer.shutdown()
+```
+Currently, this means that the user is now responsible for modifying their code to support distributed training (specifying `DistributedDataParallel` for `torch` or `MultiWorkerMirroredStrategy` for `tensorflow`) as opposed to having this be automatically handled internally. However, we have plans to provide utilities that you can use to automatically handle these recipes for you.
+
+## Training Logic
+The main change you will have to make is how you define your training logic. In Ray SGD v1, the API for defining training logic differed for `TorchTrainer` vs. `TFTrainer`, so the steps to migrate will be different for each of these.
+
+### PyTorch
+In v1, the training logic is defined through the `train_epoch` and `train_batch` methods of a `TrainingOperator` class which is passed into the `TorchTrainer`. To migrate to Ray SGD v2, there are 2 options: 
+1. If you liked having your training logic in the `TrainingOperator`, you can continue to use the `TrainingOperator` with Ray SGD v2.
+2. If you felt the `TrainingOperator` is too unnecessary and complex, or you had to customize it extensively, you can define your own training function.
+
+#### Alternative 1: Continue to use `TrainingOperator`
+You can define a training function that instantiates your `TrainingOperator` and you can call methods directly on the operator object.
+
+So instead of 
+```python
+class MyTrainingOperator(TrainingOperator):
+	...
+
+trainer = TorchTrainer(training_operator_cls=MyTrainingOperator, num_workers=4, use_gpu=True)
+
+num_epochs=10
+for _ in range(num_epochs):
+	trainer.train()
+	trainer.validate()
+
+final_model = trainer.get_model()
+```
+
+you would do
+```python
+class MyTrainingOperator(TrainingOperator):
+	...
+
+def train_func(config):
+	device = torch.device(f"cuda:{sgd.local_rank()}" if
+                  torch.cuda.is_available() else "cpu")
+	if torch.cuda.is_available():
+		torch.cuda.set_device(device)
+	
+	operator_cls = config["operator_cls"]
+	
+	# Set the args to whatever values you want.
+	training_operator = operator_cls(
+		config=config, 
+		world_rank=sgd.world_rank(), 
+		local_rank=sgd.local_rank(), 
+		is_distributed=True,
+		device=device,
+		use_gpu=True,
+		wrap_ddp=True,
+		add_dist_sampler=True 
+
+	training_operator.setup(config)
+
+	for idx in range(config["num_epochs"]):
+		train_loader = training_operator._get_train_loader()
+		# If using DistributedSampler, set the epoch here.
+		train_loader.set_epoch(idx)
+		training_operator.train_epoch(epoch_idx=idx, iter(train_loader))
+
+		validation_loader = training_operator._get_validation_loader()
+		training_operator.validate(iterator=iter(validation_loader))
+	
+	if sgd.world_rank() == 0:
+		return training_operator._get_original_models()
+	else:
+		return None
```

Comment:
nit: Should we use `save_checkpoint` API instead?

Reply:
This is also a note-to-self since it doesn't look like we do a good job describing how to use `return` in the v2 docs.

Reply:
Hmm I think for this simple case we should recommend just returning the model.